cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -92,51 +92,61 @@ func ValidateNodeRegistrationOptions(nro *kubeadm.NodeRegistrationOptions, fldPa
 	}
-	if len(c.DiscoveryFile) != 0 {
-		allErrs = append(allErrs, ValidateDiscoveryFile(c.DiscoveryFile, field.NewPath("discoveryFile"))...)
-		if len(c.TLSBootstrapToken) != 0 {
-			allErrs = append(allErrs, ValidateToken(c.TLSBootstrapToken, field.NewPath("tlsBootstrapToken"))...)
+
+	if d.BootstrapToken != nil {
+		allErrs = append(allErrs, ValidateBootstrapTokenDiscovery(d.BootstrapToken, fldPath.Child("bootstrapToken"))...)
+		allErrs = append(allErrs, ValidateToken(d.TLSBootstrapToken, fldPath.Child("tlsBootstrapToken"))...)
+	}
+
+	if d.File != nil {
+		allErrs = append(allErrs, ValidateFileDiscovery(d.File, fldPath.Child("file"))...)
+		if len(d.TLSBootstrapToken) != 0 {
+			allErrs = append(allErrs, ValidateToken(d.TLSBootstrapToken, fldPath.Child("tlsBootstrapToken"))...)
 		}
-	} else {
-		allErrs = append(allErrs, ValidateToken(c.TLSBootstrapToken, field.NewPath("tlsBootstrapToken"))...)
 	}
-	allErrs = append(allErrs, ValidateArgSelection(c, field.NewPath("discovery"))...)
-	allErrs = append(allErrs, ValidateJoinDiscoveryTokenAPIServer(c.DiscoveryTokenAPIServers, field.NewPath("discoveryTokenAPIServers"))...)
 
 	return allErrs
 }
 
-// ValidateArgSelection validates discovery related configuration and collects all encountered errors
-func ValidateArgSelection(cfg *kubeadm.JoinConfiguration, fldPath *field.Path) field.ErrorList {
+// ValidateBootstrapTokenDiscovery validates bootstrap token discovery configuration
+func ValidateBootstrapTokenDiscovery(b *kubeadm.BootstrapTokenDiscovery, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if len(cfg.DiscoveryToken) != 0 && len(cfg.DiscoveryFile) != 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath, "", "discoveryToken and discoveryFile cannot both be set"))
-	}
-	if len(cfg.DiscoveryToken) == 0 && len(cfg.DiscoveryFile) == 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath, "", "discoveryToken or discoveryFile must be set"))
-	}
-	if len(cfg.DiscoveryTokenAPIServers) < 1 && len(cfg.DiscoveryToken) != 0 {
-		allErrs = append(allErrs, field.Required(fldPath, "discoveryTokenAPIServers not set"))
-	}
-
-	if len(cfg.DiscoveryFile) != 0 && len(cfg.DiscoveryTokenCACertHashes) != 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath, "", "discoveryTokenCACertHashes cannot be used with discoveryFile"))
-	}
 
-	if len(cfg.DiscoveryFile) == 0 && len(cfg.DiscoveryToken) != 0 &&
-		len(cfg.DiscoveryTokenCACertHashes) == 0 && !cfg.DiscoveryTokenUnsafeSkipCAVerification {
-		allErrs = append(allErrs, field.Invalid(fldPath, "", "using token-based discovery without discoveryTokenCACertHashes can be unsafe. set --discovery-token-unsafe-skip-ca-verification to continue"))
+	if len(b.APIServerEndpoints) < 1 {
+		allErrs = append(allErrs, field.Required(fldPath, "APIServerEndpoints not set"))
 	}
 
 	// TODO remove once we support multiple api servers
-	if len(cfg.DiscoveryTokenAPIServers) > 1 {
+	if len(b.APIServerEndpoints) > 1 {
```

Comment:
This should go away when `APIServerEndpoints` changes from `[]string` to `string`